### PR TITLE
PP-14542: Add a new endpoint to handle Adyen 3DS redirect and submit result to Connector

### PR DESCRIPTION
**Pull request by Pippa101**
## WHAT
- added a new path for `auth3dsRequiredInAdyen`
- extracted `redirectResult` from Adyen's response to be sent to connector

### CODE REVIEW

File: app/controllers/three-d-secure.controller.js [hunk 1]
```diff
@@ -52,6 +52,13 @@ const build3dsPayload = (chargeId, req) => {
     auth3dsPayload.md = md
   }
 
+  const redirectResult = _.get(req, 'body.redirectResult')
+  if (_.isUndefined(redirectResult)) {
+    logger.info('redirectResult is not present')
```
Comment (kbottla) [inline app/controllers/three-d-secure.controller.js]: This is logged for every 3DS payment, irrespective of the payment provider, which adds noise. The most relevant place could be the `auth3dsRequiredInAdyen` method.

File: test/integration/three-d-secure.ft.test.js [hunk 2]
```diff
@@ -336,6 +336,53 @@ describe('chargeTests', function () {
       })
     })
 
+    describe('for adyen payment provider', function () {
+      it('should return the data needed for the UI when POST', function (done) {
```
Comment (kbottla) [inline test/integration/three-d-secure.ft.test.js]: Could it be renamed so it's clear? Something like below

```suggestion
      it('should return the data, including redirectResult, in the UI for POST to 3ds_required_in/adyen', function (done) {
```

File: test/integration/three-d-secure.ft.test.js [hunk 3]
```diff
@@ -336,6 +336,53 @@ describe('chargeTests', function () {
...
+            expect(res.header['set-cookie'] === undefined).to.be.true // eslint-disable-line
+            const $ = cheerio.load(res.text)
+            expect($('form[name=\'three_ds_required\'] > input[name=\'csrfToken\']').attr('value')).to.not.exist // eslint-disable-line
+            expect($('form[name=\'three_ds_required\'] > input[name=\'redirectResult\']').attr('value')).to.not.exist // eslint-disable-line
+            expect($('form[name=\'three_ds_required\']').attr('action')).to.eql(`/card_details/${chargeId}/3ds_handler`)
+          })
+          .end(done)
+      })
+
+      it('should return the data needed for the UI when GET', function (done) {
```
Comment (kbottla) [inline test/integration/three-d-secure.ft.test.js]: Same as above
```suggestion
      it('should return the data, including redirectResult, in the UI for GET to 3ds_required_in/adyen', function (done) {
```

File: test/integration/three-d-secure.ft.test.js (hunk 2 above)
Comment (kbottla) [inline test/integration/three-d-secure.ft.test.js]: Can another test be added for an empty or missing `redirectResult`?

**Review by kbottla — CHANGES_REQUESTED** [review]

File: app/controllers/three-d-secure.controller.js [hunk 4]
```diff
@@ -177,5 +182,15 @@ module.exports = {
...
       providerStatus: _.get(req, 'query.status', 'success')
     })
+  },
+  auth3dsRequiredInAdyen: (req, res) => {
+    const charge = normalise.charge(req.chargeData, req.chargeId)
+    responseRouter.response(req, res, views.AUTH_3DS_REQUIRED_IN_VIEW, {
+      threeDsHandlerUrl: routeFor('auth3dsHandler', charge.id),
+      redirectResult: _.get(req, 'query.redirectResult', '')
+    })
+    if (_.isUndefined(_.get(req, 'query.redirectResult'))) {
```
Comment (kbottla) [inline app/controllers/three-d-secure.controller.js]: Can this be moved to the top of the method so it feels more natural for `render` to be the last thing the method does?

**Review by kbottla — APPROVED** [review]
LGTM!
